### PR TITLE
Lua refactor

### PR DESCRIPTION
**Pull request by trws**
Reworking lua to allow easier substitution of the base lua implementation.  Also adding in a maintained version of luajit and re-factoring the entire stack to use a custom build_system to centralize functionlity like environment variable management and luarocks installation.

Requires: #28853

### CODE REVIEW

**Review by adamjstewart — COMMENTED** [review]
Haven't looked at this in too much detail yet. My first impression is that I agree that we need a `LuaPackage` base class for Lua extensions. I'm less certain about `LuaImplPackage` though. This seems to relate in many ways to #17475, and I think a lot of the "hackiness" I feel from this PR is due to the fact that 1) we can't directly extend a virtual package, and 2) we don't want to copy-n-paste all the activation logic to multiple Lua providers. Python and Java face these same issues, and we haven't found a good solution yet. It would be good to discuss if this solution is something that we want to use in general or if there's a better solution that can be used for all virtual providers/extensions.

Also, once we decide on a final design for this, we'll need docs in `lib/spack/docs/build_systems/luapackage.rst` and possibly build-system detection in `lib/spack/spack/cmd/create.py`.

**Comment by trws** [comment]
Yup, you're absolutely right that it relates to #17475.  If we could extend a virtual, then the extends(when) stuff could all come out.

I decided to go this way for lua specifically because:
* all of the implementations should use the same logic for:
  * extension
  * build and run environment setup
  * luarocks installation
  * providing access to luarocks

That is a whole lot of code that ought to be implemented in only one place, and at the moment the only way I know to do that within spack is to make it a build system.  I argued with myself a bit about making all of that logic part of an actual lua-lang *package* that would depend on one of the three depending on variants or something, but that seems even worse, or making it a package that gets imported by the others and used as a base, but again that didn't seem better.

What I'm trying to get to I suppose is that there are two problems here:
* no good way to extend virtuals (as in #17475)
* no good way to share common code between related packages other than build systems

The latter is something I want to find a way to do for packages like LLVM and the ROCM family.  It bugs the crap out of me that the same set of version logic is copied into each and every rocm package, so every time there's a new version every single one has to get touched to add a new number in the exact same list.  This is part of why LLVM ended up as a single package again instead of split out too, it should be several packages with *one central implementation and version*.  Not sure how best to fix that, but given the mechanisms available this seemed least hacktastic.

The thing I still really don't like is that luarocks is embedded in lua.  I say that despite being the person that did that in the first place.  It almost has to be there because that's where things should activate to, but at the same time it should be able to have a version independent of lua.

Anyway, sorry for the book, 100% agree there are things to resolve here.

**Comment by adamjstewart** [comment]
I've also thought about whether it makes sense for virtual dependencies to require actual `package.py` files so that this kind of code can be shared and we don't need to parse every single `package.py` just to determine which packages provide what.

**Comment by trws** [comment]
That might make for a nice solution.  By far the worst part of implementing this refactor was getting the two build_system files to work without import cycles in bootstrap.  Using normal package directives in build_system classes is *very* fragile.

**Comment by adamjstewart** [comment]
Pinging @alalazo @becker33 @tgamblin on this one because they usually enjoy brainstorming these kinds of conceptual design questions. See #17475 for a good description of at least half of the issue. The issue of reusing the same code for all virtual providers is kind of a second issue that we should also think about.

**Comment by trws** [comment]
Anybody have opinions: @alalazo, @becker33, @tgamblin?  It's a bit painful to be stuck with upstream lua only without something like this patch.  I don't mind re-working if we want a different mechanism, but some way forward would be really helpful.

**Comment by adamjstewart** [comment]
Took another look at this and I think my thoughts on this PR are mostly the same as when I last looked at it. Basically, `LuaPackage` looks good to me, but `LuaImplPackage` feels like a hack to work around other deficiencies in Spack.

I think the "right" thing to do is to fix #17475. Of course, this is easier said than done, and might take a while to get right. We might have to merge something like this PR while we wait.

Another design question I have is what actually belongs in `lib/spack/spack/build_systems`? Originally it was only build systems, then we started adding mixins for things like Cuda and ROCm. Obviously mixins are not build systems, and `LuaImplPackage` is even more different. I think `LuaImplPackage` may be better to store in the `package.py` of one of the Lua providers. For example, we do something very similar with `py-pillow` and `py-pillow-simd`. Thoughts?

File: lib/spack/spack/build_systems/lua.py
Comment (alalazo) [inline lib/spack/spack/build_systems/lua.py]: Is this a typo:
```suggestion
        return Executable(join_path(self.spec['lua-lang'].prefix.bin, 'lua'))
```
?

**Review by alalazo — CHANGES_REQUESTED** [review]
I don't mind this implementation while we look at a better refactoring. For the `LuaImplPackage`, is there one of the three implementations that is a de-facto reference? If so we can put that class in the corresponding package and have other implementations:
```python
from spack.pkg.builtin.*** import LuaImplPackage
```

Minor: can you fix style issues, so we can get other checks running too?

File: lib/spack/spack/build_systems/lua.py
Comment (trws) [reply lib/spack/spack/build_systems/lua.py]: Yes, that's a typo, thanks!

**Comment by trws** [comment]
Ok, re-factored to put `LuaImplPackage` in the lua package.  Fixing a couple of style checking issues, and repaired the properties.  I'm going to go over and rebase the neovim fixup on top of this to get some better downstream testing.

**Comment by adamjstewart** [comment]
> Also, once we decide on a final design for this, we'll need docs in `lib/spack/docs/build_systems/luapackage.rst` and possibly build-system detection in `lib/spack/spack/cmd/create.py`.

Ping on this. At this point you're probably the only Spack developer who has a good understanding of Lua's build system.

**Comment by trws** [comment]
Ah, good point, I'll do that shortly.

**Comment by trws** [comment]
The create integration is in and working.  I don't often work with rst, so I'm not as confident the docs are the way they need to be, but they are there.

File: lib/spack/docs/build_systems/luapackage.rst
Comment (adamjstewart) [inline lib/spack/docs/build_systems/luapackage.rst]: This first section still references Meson

File: lib/spack/docs/build_systems/luapackage.rst
Comment (adamjstewart) [inline lib/spack/docs/build_systems/luapackage.rst]: Probably better to use a glob `*.rockspec`

File: lib/spack/docs/build_systems/luapackage.rst
Comment (adamjstewart) [inline lib/spack/docs/build_systems/luapackage.rst:49]: Can you add a 1-line example of this? This helps people know what text to search for in that file

File: lib/spack/spack/build_systems/lua.py
Comment (adamjstewart) [inline lib/spack/spack/build_systems/lua.py]: Where is LuaRocksPackage?

File: lib/spack/spack/build_systems/lua.py
Comment (adamjstewart) [inline lib/spack/spack/build_systems/lua.py]: Probably better to make this a function like `cmake_args` or `configure_args`

File: lib/spack/spack/cmd/create.py
Comment (adamjstewart) [inline lib/spack/spack/cmd/create.py:696]: You can also update `lib/spack/spack/test/build_system_guess.py` to add tests for this.

File: lib/spack/spack/build_systems/lua.py
Comment (trws) [reply lib/spack/spack/build_systems/lua.py]: Dang, that's a holdout from a refactor.  I initially had luapackage as a specialized MakefilePackage, but it just isn't consistent enough to be useful, so I got rid of it and promoted LuaRocksPackage to LuaPackage.

File: lib/spack/spack/build_systems/lua.py
Comment (trws) [reply lib/spack/spack/build_systems/lua.py]: Fair, updated.

File: lib/spack/docs/build_systems/luapackage.rst
Comment (trws) [reply lib/spack/docs/build_systems/luapackage.rst:49]: Sure, I'll use this one:

```lua
dependencies = {
   "lua >= 5.1",
}
```

File: lib/spack/spack/cmd/create.py
Comment (trws) [reply lib/spack/spack/cmd/create.py:696]: Thanks for that, added.

File: lib/spack/docs/build_systems/luapackage.rst
Comment (adamjstewart) [inline lib/spack/docs/build_systems/luapackage.rst]:
```suggestion
a dependency. The LuaPackage class adds appropriate dependencies on a Lua
```

File: lib/spack/docs/build_systems/luapackage.rst
Comment (adamjstewart) [inline lib/spack/docs/build_systems/luapackage.rst:59]: What would be the equivalent syntax for Spack? Does we need to specify it for multiple possible providers at the moment?

File: lib/spack/spack/build_systems/lua.py
Comment (adamjstewart) [inline lib/spack/spack/build_systems/lua.py]: Let's avoid star imports and specify depends_on, extends explicitly

File: lib/spack/spack/build_systems/lua.py
Comment (adamjstewart) [inline lib/spack/spack/build_systems/lua.py]: I think this line can be deleted

File: lib/spack/spack/build_systems/lua.py
Comment (adamjstewart) [inline lib/spack/spack/build_systems/lua.py]:
```suggestion
        """Override this to preprocess source before building with luarocks"""
```

File: lib/spack/spack/build_systems/lua.py
Comment (adamjstewart) [inline lib/spack/spack/build_systems/lua.py]:
```suggestion
        return Executable(self.spec['lua-lang'].prefix.bin.lua)
```

File: lib/spack/spack/build_systems/lua.py
Comment (adamjstewart) [inline lib/spack/spack/build_systems/lua.py]:
```suggestion
        return Executable(self.spec['lua-lang'].prefix.bin.luarocks)
```

File: lib/spack/spack/build_systems/lua.py
Comment (adamjstewart) [inline lib/spack/spack/build_systems/lua.py]: Isn't this already a list?

File: lib/spack/docs/build_systems/luapackage.rst
Comment (adamjstewart) [inline lib/spack/docs/build_systems/luapackage.rst]: Are there any examples of packages that use this (or will need this) phase? Or is this just hypothetical?

File: lib/spack/spack/build_systems/lua.py
Comment (adamjstewart) [inline lib/spack/spack/build_systems/lua.py]: I think these can be removed now

File: lib/spack/spack/cmd/create.py
Comment (adamjstewart) [inline lib/spack/spack/cmd/create.py]:
```suggestion
        # FIXME: Add arguments to `luarocks make` other than rockspec path
```

File: var/spack/repos/builtin/packages/lua-lpeg/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/lua-lpeg/package.py]: Does the default install method not work?

File: var/spack/repos/builtin/packages/lua/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/lua/package.py]: This shouldn't be needed in packages, these are all already included

File: var/spack/repos/builtin/packages/lua/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/lua/package.py:10]: I think this is something we've been meaning to add to all packages at some point, not sure if we ever made any progress on that

File: var/spack/repos/builtin/packages/lua/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/lua/package.py]:
```suggestion
        return Executable(self.spec.prefix.bin.lua)
```

File: var/spack/repos/builtin/packages/lua/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/lua/package.py]:
```suggestion
        return Executable(self.spec.prefix.bin.luarocks)
```

File: var/spack/repos/builtin/packages/lua/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/lua/package.py]: I think this line needs to be indented for the double colon above to work. Not that we generate docs for packages anyway.

File: var/spack/repos/builtin/packages/lua/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/lua/package.py]:
```suggestion
        module.lua = Executable(self.spec.prefix.bin.lua)
```

File: var/spack/repos/builtin/packages/lua/package.py
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/lua/package.py]:
```suggestion
        module.luarocks = Executable(self.spec.prefix.bin.luarocks)
```

File: lib/spack/docs/build_systems/luapackage.rst
Comment (trws) [reply lib/spack/docs/build_systems/luapackage.rst]: I had one that needed it, but reworked it using a newer luarocks dependency.  There are a *lot* of old largely unmaintained but critical lua packages around, I expect it to come up every so often.

File: lib/spack/docs/build_systems/luapackage.rst
Comment (trws) [reply lib/spack/docs/build_systems/luapackage.rst:59]: I suppose it would be something like:

```python
depends_on('lua@5.1:5.1.99', when='^lua')
```

It's only one because it's 5.1, and all the luajit variants are 5.1 variants.  If it were 5.3, only lua would be valid, so it would become:

```python
depends_on('lua@5.3')
```

That's actually a bit of an oddity with all this.  I considered putting a version on lua-lang, from a logical standpoint it should probably be there, but there's only one lua implementation that provides anything other than 5.1.

File: lib/spack/spack/build_systems/lua.py
Comment (trws) [reply lib/spack/spack/build_systems/lua.py]: That was there to copy the list more than convert it, but now that it's returned from a function it's no longer necessary.  Removed.

File: var/spack/repos/builtin/packages/lua/package.py
Comment (trws) [reply var/spack/repos/builtin/packages/lua/package.py:10]: Yeah, it's on my list.  That and getting things out of the module, and into package instances or similar so we can actually typecheck these darned things would make me really happy.  It keeps getting pushed back by functional stuff for build though.  At least putting it in as I go gets most of the type checking failures out for the ones I'm touching.

File: var/spack/repos/builtin/packages/lua/package.py
Comment (trws) [reply var/spack/repos/builtin/packages/lua/package.py]: True, same for the filesystem ones.  They came in from being needed in the build_system version, but are now gone.

File: var/spack/repos/builtin/packages/lua-lpeg/package.py
Comment (trws) [reply var/spack/repos/builtin/packages/lua-lpeg/package.py]: Thanks for the reminder.  This version of the package would not have worked, no, because lpeg was using the actual rock archive from the luarocks distro rather than a source package.  Basically a luarocks wheel.  I'm reworking it to use the source instead then the default works just fine.

File: var/spack/repos/builtin/packages/lua/package.py
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/lua/package.py:10]: If we could get it in before 0.18 (sometime in the next week) that would be amazing. It's hard to add after because it becomes difficult to backport.

File: lib/spack/docs/build_systems/luapackage.rst
Comment (trws) [reply lib/spack/docs/build_systems/luapackage.rst]: This is actually back in active use, the lua-bitlib package needs it to fix the `luaL_reg` calls to be capitalized correctly.

File: var/spack/repos/builtin/packages/lua-lpeg/package.py
Comment (trws) [reply var/spack/repos/builtin/packages/lua-lpeg/package.py]: Spack now supports the rock archives directly, and this is reworked to use it.

File: lib/spack/docs/build_systems/luapackage.rst
Comment (trws) [reply lib/spack/docs/build_systems/luapackage.rst:59]: I added some stuff to the docs, but in principle this should now be `depends_on('lua-lang@5.1:')`, as long as the package doesn't care what implementation it gets.

File: var/spack/repos/builtin/packages/lua/package.py
Comment (trws) [reply var/spack/repos/builtin/packages/lua/package.py:10]: I can try to dig out my old rope script and just do it, it basically can't break anything, so I think it's mainly a question of getting it in and writing something to make sure every new package has it.  Maybe add an isort rule? Will look into it.

**Comment by trws** [comment]
Ok, that took a *lot* more changes than I had been hoping, but I updated every lua package I could find that was manually using luarocks, and extended the luarocks support to correctly handle cross-tree dependencies and source rock archives, along with new support in create and fetch to handle that.  😮‍💨 In the end, it seems to work, across lua implementations and versions and packages.  Also fixed missing dependencies we were accidentally fetching from the internet previously.

**Comment by trws** [comment]
Thanks to @tgamblin for some help spelunking the ASP, the last commit fixes concretizing to handle `when='variant@version'` allowing the luaposix bit32 dependency to be described accurately.  Assuming this passes, it should be ready for a re-review.

**Comment by adamjstewart** [comment]
One alternative to the `unpack` phase of `LuaPackage` would be to support directly extracting `*.rock` files in `lib/spack/spack/util/compression.py`. Advantage would be one less phase and chdir and removes the need for `extract=False` in the packages. Disadvantage would be needing to search for `luarocks` in the `PATH` via `which` in this location, but that shouldn't be an issue since we'll always have a dep on lua.

This PR is starting to get big. We may want to split this into multiple PRs if it gets much bigger.

**Comment by trws** [comment]
I actually tried to implement it that way first, but ran into issues with that solution because the unpack makes a two-level directory result that doesn't get handled properly by the directory identification machinery, and honestly I don't even know where to look to fix that.  If you can point me in the general direction, I wouldn't mind doing it, but it was becoming too much work, hence the rework into a phase.

**Review by adamjstewart — DISMISSED** [review]
I'm happy with the current state of this. We should probably get a review from at least one more person before merging. Great work!

**Comment by alalazo** [comment]
With respect to this: https://github.com/spack/spack/pull/28854#issuecomment-1112775790 did we check if there's any impact on concretization performance?

File: lib/spack/spack/build_systems/lua.py
Comment (alalazo) [inline lib/spack/spack/build_systems/lua.py]: I think this should work too and is a bit more readable:
```suggestion
    with when('^lua-luajit'):
        extends('lua-luajit')
        depends_on('luajit')
        depends_on('lua-luajit+lualinks')
    
    with when('^lua-luajit-openresty'):
        extends('lua-luajit-openresty')
        depends_on('luajit')
        depends_on('lua-luajit-openresty+lualinks')
```

File: lib/spack/spack/build_systems/lua.py
Comment (alalazo) [inline lib/spack/spack/build_systems/lua.py:42]: Can we store the directory and then use the context manager `working_dir` where appropriate? That should restore the current working directory on exit from the context.

File: lib/spack/spack/build_systems/lua.py
Comment (alalazo) [inline lib/spack/spack/build_systems/lua.py:81]: Why are we adding an empty phase with a hook that runs before that (`_generate_luarocks_config`)? Wouldn't make more sense to have the hook be the phase and eventually have people `run_after` that?

File: lib/spack/spack/package.py
Comment (alalazo) [inline lib/spack/spack/package.py]: This introduces a bug. It's weird, but the argument is `deptype`, see #29397

**Review by alalazo — CHANGES_REQUESTED** [review]

File: lib/spack/spack/build_systems/lua.py
Comment (trws) [reply lib/spack/spack/build_systems/lua.py:81]: The config generation could be a phase. I think I did it this way so that if the user uses luarocks for something during preprocess they'll get the config, they could edit the generated config in preprocess, but they can't accidentally prevent it from being run by overloading the phase.  The empty phase is there because there are a number of packages that need editing or patching to compile, serves a function much like `edit` in MakefilePackage.

File: lib/spack/spack/build_systems/lua.py
Comment (trws) [reply lib/spack/spack/build_systems/lua.py:42]: We can certainly store the directory, but I've had zero luck figuring out how to use `with working_dir` in an un-scoped manner like this needs.  Is it possible to set the `build_directory` dynamically in an early phase and tell spack to re-evaluate it or something? I agree that I'm unfond of this but hadn't come up with a better solve.

File: lib/spack/spack/package.py
Comment (trws) [reply lib/spack/spack/package.py]: I wonder why I did this... I have a vague memory that it actually fixed something, which is disturbing in its own way, I'll look into it.

File: lib/spack/spack/build_systems/lua.py
Comment (alalazo) [inline lib/spack/spack/build_systems/lua.py:11]:
```suggestion
from spack.directives import depends_on, extends
from spack.multimethod import when
```

File: lib/spack/spack/build_systems/lua.py
Comment (alalazo) [inline lib/spack/spack/build_systems/lua.py]:
```suggestion

```

**Review by alalazo — COMMENTED** [review]
A few fixes for style

File: lib/spack/spack/build_systems/lua.py
Comment (alalazo) [reply lib/spack/spack/build_systems/lua.py:81]: > The config generation could be a phase.

Yeah, let's make it a phase on its own, I'd be good with that.

File: lib/spack/spack/build_systems/lua.py
Comment (alalazo) [reply lib/spack/spack/build_systems/lua.py:42]: If that is hard to do, I would not block merging this PR. Can you leave a `TODO` comment though or something like that?

Comment (trws) [reply lib/spack/spack/build_systems/lua.py:42]: Sure, adding one now.

**Comment by trws** [comment]
> With respect to this: [#28854 (comment)](https://github.com/spack/spack/pull/28854#issuecomment-1112775790) did we check if there's any impact on concretization performance?

I did some testing, both on the full PR and on just the changes in the ASP.  The change in the ASP was basically nil, but with the full PR there's a small but noticeable increase in solve time.  I'm guessing it's because of the additional virtual.

**Comment by trws** [comment]
@spackbot run pipeline

**Review by alalazo — APPROVED** [review]
Thanks!

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**
